cabal-extract/Main.hs

```diff
@@ -0,0 +1,261 @@
+    rejectedNamePattern = "^\\..*|^dist-newstyle$"
+    allowedNamePattern :: String
+    allowedNamePattern = ".*\\.cabal"
+
+parseCabalFile :: FilePath -> IO GenericPackageDescription
+parseCabalFile = readGenericPackageDescription deafening
+
+convertLibrary :: Library -> Component
+convertLibrary lib =
+  Component
+    { componentName = case libName lib of
+        LMainLibName ->
+          "library"
+        LSubLibName ucn ->
+          unUnqualComponentName ucn,
+      componentDependencies = targetBuildDepends (libBuildInfo lib),
+      componentReverseDependencies = []
+    }
+
+convertExecutable :: Executable -> Component
+convertExecutable exe =
+  Component
+    { componentName = unUnqualComponentName (exeName exe),
+      componentDependencies = targetBuildDepends (buildInfo exe),
+      componentReverseDependencies = []
+    }
+
+convertTest :: TestSuite -> Component
+convertTest t =
+  Component
+    { componentName = unUnqualComponentName (testName t),
+      componentDependencies = targetBuildDepends (testBuildInfo t),
+      componentReverseDependencies = []
+    }
+
+convertBenchmark :: Benchmark -> Component
+convertBenchmark b =
+  Component
+    { componentName = unUnqualComponentName (benchmarkName b),
+      componentDependencies = targetBuildDepends (benchmarkBuildInfo b),
+      componentReverseDependencies = []
+    }
+
+convertPackage :: PackageDescription -> Package
+convertPackage pd =
+  Package
+    { packageName =
+        unPackageName (pkgName (package pd)),
+      packageVersion =
+        prettyShow (pkgVersion (package pd)),
+      packageLibraries =
+        maybe [] (pure . convertLibrary) (library pd)
+          ++ map convertLibrary (subLibraries pd),
+      packageExecutables =
+        map convertExecutable (executables pd),
+      packageTests =
+        map convertTest (testSuites pd),
+      packageBenchmarks =
+        map convertBenchmark (benchmarks pd)
+    }
+
+depToString :: Dependency -> String
+depToString dep =
+  intercalate "," $
+    map fmt (NES.toList (depLibraries dep))
+  where
+    pkg = unPackageName (depPkgName dep)
+    fmt LMainLibName = pkg
+    fmt (LSubLibName subname) = pkg ++ ":" ++ unUnqualComponentName subname
+
+allComponents :: Package -> [Component]
+allComponents p =
+  packageLibraries p
+    ++ packageExecutables p
+    ++ packageTests p
+    ++ packageBenchmarks p
+
+innerComponent :: [Package] -> [String]
+innerComponent =
+  concatMap
+    ( \pkg ->
+        map
+          ( \comp -> case componentName comp of
+              "library" -> packageName pkg
+              name -> packageName pkg ++ ":" ++ name
+          )
+          (allComponents pkg)
+    )
+
+trimComponent :: Set.Set String -> Component -> Component
+trimComponent allowed comp =
+  comp
+    { componentDependencies =
+        filter
+          (\dep -> depToString dep `Set.member` allowed)
+          (componentDependencies comp)
+    }
+
+trim :: Package -> [String] -> Package
```

Comment:
The name of this function is not clear enough to give proper context on what it actually does
```suggestion
keepWhitelistedDependencies :: Package -> [String] -> Package
```